File: php/tests/GildedRoseTest.php
<?php

namespace Tests;

use GildedRose\GildedRose;
use GildedRose\Item;
use PHPUnit\Framework\TestCase;

class GildedRoseTest extends TestCase
{
    public function testNormalItem_BeforeSellDate(): void
    {
        $items = [new Item('randomstring', 5, 10)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(4, $items[0]->days_remaining);
        $this->assertSame(9, $items[0]->quality);
    }

    public function testNormalItem_WithMinQuality(): void
    {
        $items = [new Item('randomstring', 5, 0)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(4, $items[0]->days_remaining);
        $this->assertSame(0, $items[0]->quality);
    }

    public function testNormalItem_OnSellDate(): void
    {
        $items = [new Item('randomstring', 0, 10)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(-1, $items[0]->days_remaining);
        $this->assertSame(8, $items[0]->quality);
    }

    public function testNormalItem_OnSellDateWithMinQuality(): void
    {
        $items = [new Item('randomstring', 0, 0)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(-1, $items[0]->days_remaining);
        $this->assertSame(0, $items[0]->quality);
    }

    public function testNormalItem_OnSellDateNearMinQuality(): void
    {
        $items = [new Item('randomstring', 0, 1)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(-1, $items[0]->days_remaining);
        $this->assertSame(0, $items[0]->quality);
    }

    public function testNormalItem_AfterSellDate(): void
    {
        $items = [new Item('randomstring', -10, 10)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(-11, $items[0]->days_remaining);
        $this->assertSame(8, $items[0]->quality);
    }

    public function testNormalItem_AfterSellDateWithMinQuality(): void
    {
        $items = [new Item('randomstring', -10, 0)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(-11, $items[0]->days_remaining);
        $this->assertSame(0, $items[0]->quality);
    }

    public function testNormalItem_AfterSellDateNearMinQuality(): void
    {
        $items = [new Item('randomstring', -10, 1)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(-11, $items[0]->days_remaining);
        $this->assertSame(0, $items[0]->quality);
    }

    public function testAgedCheddar_BeforeSellDate(): void
    {
        $items = [new Item('Aged Cheddar', 5, 10)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(4, $items[0]->days_remaining);
        $this->assertSame(11, $items[0]->quality);
    }

    public function testAgedCheddar_WithMaxQuality(): void
    {
        $items = [new Item('Aged Cheddar', 5, 50)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(4, $items[0]->days_remaining);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testAgedCheddar_OnSellDate(): void
    {
        $items = [new Item('Aged Cheddar', 0, 10)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(-1, $items[0]->days_remaining);
        $this->assertSame(12, $items[0]->quality);
    }

    public function testAgedCheddar_OnSellDateWithMaxQuality(): void
    {
        $items = [new Item('Aged Cheddar', 0, 50)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(-1, $items[0]->days_remaining);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testAgedCheddar_OnSellDateNearMaxQuality(): void
    {
        $items = [new Item('Aged Cheddar', 0, 49)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(-1, $items[0]->days_remaining);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testAgedCheddar_AfterSellDate(): void
    {
        $items = [new Item('Aged Cheddar', -10, 10)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(-11, $items[0]->days_remaining);
        $this->assertSame(12, $items[0]->quality);
    }

    public function testAgedCheddar_AfterSellDateWithMaxQuality(): void
    {
        $items = [new Item('Aged Cheddar', -10, 50)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(-11, $items[0]->days_remaining);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testAgedCheddar_AfterSellDateNearMaxQuality(): void
    {
        $items = [new Item('Aged Cheddar', -10, 49)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(-11, $items[0]->days_remaining);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testHammer_BeforeSellDate(): void
    {
        $items = [new Item('Hammer', 5, 40)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(5, $items[0]->days_remaining);
        $this->assertSame(40, $items[0]->quality);
    }

    public function testHammer_OnSellDate(): void
    {
        $items = [new Item('Hammer', 0, 40)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(0, $items[0]->days_remaining);
        $this->assertSame(40, $items[0]->quality);
    }

    public function testHammer_AfterSellDate(): void
    {
        $items = [new Item('Hammer', -10, 40)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(-10, $items[0]->days_remaining);
        $this->assertSame(40, $items[0]->quality);
    }

    public function testConcertTickets_LongBeforeSellDate(): void
    {
        $items = [new Item('Concert Tickets', 11, 10)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(10, $items[0]->days_remaining);
        $this->assertSame(11, $items[0]->quality);
    }

    public function testConcertTickets_LongBeforeSellDateAtMaxQuality(): void
    {
        $items = [new Item('Concert Tickets', 11, 50)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(10, $items[0]->days_remaining);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testConcertTickets_MediumCloseToSellDateUpperBound(): void
    {
        $items = [new Item('Concert Tickets', 10, 10)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(9, $items[0]->days_remaining);
        $this->assertSame(12, $items[0]->quality);
    }

    public function testConcertTickets_MediumCloseToSellDateUpperBoundAtMaxQuality(): void
    {
        $items = [new Item('Concert Tickets', 10, 50)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(9, $items[0]->days_remaining);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testConcertTickets_MediumCloseToSellDateUpperBoundNearMaxQuality(): void
    {
        $items = [new Item('Concert Tickets', 10, 49)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(9, $items[0]->days_remaining);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testConcertTickets_MediumCloseToSellDateLowerBound(): void
    {
        $items = [new Item('Concert Tickets', 6, 10)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(5, $items[0]->days_remaining);
        $this->assertSame(12, $items[0]->quality);
    }

    public function testConcertTickets_MediumCloseToSellDateLowerBoundAtMaxQuality(): void
    {
        $items = [new Item('Concert Tickets', 6, 50)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(5, $items[0]->days_remaining);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testConcertTickets_MediumCloseToSellDateLowerBoundNearMaxQuality(): void
    {
        $items = [new Item('Concert Tickets', 6, 49)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(5, $items[0]->days_remaining);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testConcertTickets_VeryCloseToSellDateUpperBound(): void
    {
        $items = [new Item('Concert Tickets', 5, 10)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(4, $items[0]->days_remaining);
        $this->assertSame(13, $items[0]->quality);
    }

    public function testConcertTickets_VeryCloseToSellDateUpperBoundAtMaxQuality(): void
    {
        $items = [new Item('Concert Tickets', 5, 50)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(4, $items[0]->days_remaining);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testConcertTickets_VeryCloseToSellDateUpperBoundNearMaxQuality(): void
    {
        $items = [new Item('Concert Tickets', 5, 48)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(4, $items[0]->days_remaining);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testConcertTickets_VeryCloseToSellDateLowerBound(): void
    {
        $items = [new Item('Concert Tickets', 1, 10)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(0, $items[0]->days_remaining);
        $this->assertSame(13, $items[0]->quality);
    }

    public function testConcertTickets_VeryCloseToSellDateLowerBoundAtMaxQuality(): void
    {
        $items = [new Item('Concert Tickets', 1, 50)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(0, $items[0]->days_remaining);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testConcertTickets_VeryCloseToSellDateLowerBoundNearMaxQuality(): void
    {
        $items = [new Item('Concert Tickets', 1, 48)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(0, $items[0]->days_remaining);
        $this->assertSame(50, $items[0]->quality);
    }

    public function testConcertTickets_OnSellDate(): void
    {
        $items = [new Item('Concert Tickets', 0, 10)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(-1, $items[0]->days_remaining);
        $this->assertSame(0, $items[0]->quality);
    }

    public function testConcertTickets_AfterSellDate(): void
    {
        $items = [new Item('Concert Tickets', -10, 10)];
        $gildedRose = new GildedRose();

        $gildedRose->processEndOfDay($items);

        $this->assertSame(-11, $items[0]->days_remaining);
        $this->assertSame(0, $items[0]->quality);
    }
}

File: go/gildedrose_test.go
package gildedrose

import (
	"testing"

	"gotest.tools/v3/assert"
)

func TestNormalItemBeforeSellDate(t *testing.T) {
	items := []*Item{{"randomstring", 5, 10}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 4)
	assert.Equal(t, items[0].quality, 9)
}

func TestNormalItemWithMinQuality(t *testing.T) {
	items := []*Item{{"randomstring", 5, 0}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 4)
	assert.Equal(t, items[0].quality, 0)
}

func TestNormalItemOnSellDate(t *testing.T) {
	items := []*Item{{"randomstring", 0, 10}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, -1)
	assert.Equal(t, items[0].quality, 8)
}

func TestNormalItemOnSellDateWithMinQuality(t *testing.T) {
	items := []*Item{{"randomstring", 0, 0}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, -1)
	assert.Equal(t, items[0].quality, 0)
}

func TestNormalItemOnSellDateNearMinQuality(t *testing.T) {
	items := []*Item{{"randomstring", 0, 1}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, -1)
	assert.Equal(t, items[0].quality, 0)
}

func TestNormalItemAfterSellDate(t *testing.T) {
	items := []*Item{{"randomstring", -10, 10}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, -11)
	assert.Equal(t, items[0].quality, 8)
}

func TestNormalItemAfterSellDateWithMinQuality(t *testing.T) {
	items := []*Item{{"randomstring", -10, 0}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, -11)
	assert.Equal(t, items[0].quality, 0)
}

func TestNormalItemAfterSellDateNearMinQuality(t *testing.T) {
	items := []*Item{{"randomstring", -10, 1}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, -11)
	assert.Equal(t, items[0].quality, 0)
}

func TestAgedCheddarBeforeSellDate(t *testing.T) {
	items := []*Item{{"Aged Cheddar", 5, 10}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 4)
	assert.Equal(t, items[0].quality, 11)
}

func TestAgedCheddarWithMaxQuality(t *testing.T) {
	items := []*Item{{"Aged Cheddar", 5, 50}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 4)
	assert.Equal(t, items[0].quality, 50)
}

func TestAgedCheddarOnSellDate(t *testing.T) {
	items := []*Item{{"Aged Cheddar", 0, 10}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, -1)
	assert.Equal(t, items[0].quality, 12)
}

func TestAgedCheddarOnSellDateWithMaxQuality(t *testing.T) {
	items := []*Item{{"Aged Cheddar", 0, 50}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, -1)
	assert.Equal(t, items[0].quality, 50)
}

func TestAgedCheddarOnSellDateNearMaxQuality(t *testing.T) {
	items := []*Item{{"Aged Cheddar", 0, 49}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, -1)
	assert.Equal(t, items[0].quality, 50)
}

func TestAgedCheddarAfterSellDate(t *testing.T) {
	items := []*Item{{"Aged Cheddar", -10, 10}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, -11)
	assert.Equal(t, items[0].quality, 12)
}

func TestAgedCheddarAfterSellDateWithMaxQuality(t *testing.T) {
	items := []*Item{{"Aged Cheddar", -10, 50}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, -11)
	assert.Equal(t, items[0].quality, 50)
}

func TestAgedCheddarAfterSellDateNearMaxQuality(t *testing.T) {
	items := []*Item{{"Aged Cheddar", -10, 49}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, -11)
	assert.Equal(t, items[0].quality, 50)
}

func TestHammerBeforeSellDate(t *testing.T) {
	items := []*Item{{"Hammer", 5, 40}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 5)
	assert.Equal(t, items[0].quality, 40)
}

func TestHammerOnSellDate(t *testing.T) {
	items := []*Item{{"Hammer", 0, 40}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 0)
	assert.Equal(t, items[0].quality, 40)
}

func TestHammerAfterSellDate(t *testing.T) {
	items := []*Item{{"Hammer", -10, 40}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, -10)
	assert.Equal(t, items[0].quality, 40)
}

func TestConcertTicketsLongBeforeSellDate(t *testing.T) {
	items := []*Item{{"Concert Tickets", 11, 10}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 10)
	assert.Equal(t, items[0].quality, 11)
}

func TestConcertTicketsLongBeforeSellDateAtMaxQuality(t *testing.T) {
	items := []*Item{{"Concert Tickets", 11, 50}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 10)
	assert.Equal(t, items[0].quality, 50)
}

func TestConcertTicketsMediumCloseToSellDateUpperBound(t *testing.T) {
	items := []*Item{{"Concert Tickets", 10, 10}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 9)
	assert.Equal(t, items[0].quality, 12)
}

func TestConcertTicketsMediumCloseToSellDateUpperBoundAtMaxQuality(t *testing.T) {
	items := []*Item{{"Concert Tickets", 10, 50}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 9)
	assert.Equal(t, items[0].quality, 50)
}

func TestConcertTicketsMediumCloseToSellDateUpperBoundNearMaxQuality(t *testing.T) {
	items := []*Item{{"Concert Tickets", 10, 49}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 9)
	assert.Equal(t, items[0].quality, 50)
}

func TestConcertTicketsMediumCloseToSellDateLowerBound(t *testing.T) {
	items := []*Item{{"Concert Tickets", 6, 10}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 5)
	assert.Equal(t, items[0].quality, 12)
}

func TestConcertTicketsMediumCloseToSellDateLowerBoundAtMaxQuality(t *testing.T) {
	items := []*Item{{"Concert Tickets", 6, 50}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 5)
	assert.Equal(t, items[0].quality, 50)
}

func TestConcertTicketsMediumCloseToSellDateLowerBoundNearMaxQuality(t *testing.T) {
	items := []*Item{{"Concert Tickets", 6, 49}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 5)
	assert.Equal(t, items[0].quality, 50)
}

func TestConcertTicketsVeryCloseToSellDateUpperBound(t *testing.T) {
	items := []*Item{{"Concert Tickets", 5, 10}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 4)
	assert.Equal(t, items[0].quality, 13)
}

func TestConcertTicketsVeryCloseToSellDateUpperBoundAtMaxQuality(t *testing.T) {
	items := []*Item{{"Concert Tickets", 5, 50}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 4)
	assert.Equal(t, items[0].quality, 50)
}

func TestConcertTicketsVeryCloseToSellDateUpperBoundNearMaxQuality(t *testing.T) {
	items := []*Item{{"Concert Tickets", 5, 48}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 4)
	assert.Equal(t, items[0].quality, 50)
}

func TestConcertTicketsVeryCloseToSellDateLowerBound(t *testing.T) {
	items := []*Item{{"Concert Tickets", 1, 10}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 0)
	assert.Equal(t, items[0].quality, 13)
}

func TestConcertTicketsVeryCloseToSellDateLowerBoundAtMaxQuality(t *testing.T) {
	items := []*Item{{"Concert Tickets", 1, 50}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 0)
	assert.Equal(t, items[0].quality, 50)
}

func TestConcertTicketsVeryCloseToSellDateLowerBoundNearMaxQuality(t *testing.T) {
	items := []*Item{{"Concert Tickets", 1, 48}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, 0)
	assert.Equal(t, items[0].quality, 50)
}

func TestConcertTicketsOnSellDate(t *testing.T) {
	items := []*Item{{"Concert Tickets", 0, 10}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, -1)
	assert.Equal(t, items[0].quality, 0)
}

func TestConcertTicketsAfterSellDate(t *testing.T) {
	items := []*Item{{"Concert Tickets", -10, 10}}
	GildedRose{}.ProcessEndOfDay(items)
	assert.Equal(t, items[0].daysRemaining, -11)
	assert.Equal(t, items[0].quality, 0)
}
